### PR TITLE
support offline consolidate and reshard fsdp checkpoints

### PR DESCRIPTION
**Pull request by hanwen-sun**
Support to consolidate and reshard xls fsdp checkpoints offline.

the time and memory need:

  | load | consolidate | reshard | save | Total | memory
-- | -- | -- | -- | -- | -- | --
llama3-8b model fsdp=8 reshard = 1 | 7s | 1s |   | 20s | 28s |  
llama3-8b optimizer fsdp=8 reshard = 1 | 16s | 2s |   | 40s | 58s |  
**llama3-8b consolidate 8** |   |   |   |   | **1.5min** |  
llama3-8b model fsdp=8 reshard = 4 | 7s | 1s | 2s | 7s | 17s |  
llama3-8b optimizer fsdp=8 reshard = 4 | 16s | 2s | 4s | 13s | 35s |  
**llama3-8b reshard 8 - 4** |   |   |   |   | **52s** |  
llama3-70b model fsdp=32 reshard = 1 | 50s | 8s |   | 180s | 238s |  
llama3-70b optimizer fsdp=32 reshard = 1 | 100s | 18s |   | 480s  | 598s | 530G
**llama3-70b consolidate 32** |   |   |   |   | **14 min** |  
llama3-70b model fsdp=32 reshard = 16 | 50 | 8s | 15s | 50s | 123s |  
llama3-70b optimizer fsdp=32 reshard = 16 | 100s | 18s | 36s | 100s | 254s | 530G
**llama3-70b reshard 32 - 16** |   |   |   |   | **6.3min** |  

</article>

### CODE REVIEW

File: torchacc/dist/state_dict_utils.py [hunk 1]
```diff
@@ -0,0 +1,567 @@
+from collections import OrderedDict
```
Comment (anw90) [inline torchacc/dist/state_dict_utils.py]: Use isort to sort the import statements.

File: torchacc/utils/consolidate_and_reshard_ckpts.py [hunk 2]
```diff
@@ -0,0 +1,73 @@
+from argparse import ArgumentParser
```
Comment (anw90) [inline torchacc/utils/consolidate_and_reshard_ckpts.py:1]: Add this script to setup.py so users can run consolidate_and_reshape_ckpts directly from the command line after installing torchacc.

File: torchacc/utils/consolidate_and_reshard_ckpts.py [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+from torchacc.dist.state_dict_utils import consolidate_and_reshard_model_dict, consolidate_and_reshard_optim_dict
+
+MODEL_NAME_PATTERN = "rank*-of-*-model.pth"
```
Comment (anw90) [inline torchacc/utils/consolidate_and_reshard_ckpts.py]: maybe we could make these two patterns as job arguments?

File: torchacc/utils/consolidate_and_reshard_ckpts.py [hunk 4]
```diff
@@ -0,0 +1,73 @@
+from argparse import ArgumentParser
+from torchacc.dist.state_dict_utils import consolidate_and_reshard_model_dict, consolidate_and_reshard_optim_dict
```
Comment (anw90) [inline torchacc/utils/consolidate_and_reshard_ckpts.py]: Add FSDP information to this file name to clarify its usage scenario.

File: torchacc/utils/consolidate_and_reshard_ckpts.py [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+        type=str,
+        default="",
+        help=(
+            f"The name pattern of the XLA FSDP checkpoint files to be consolidated. "
+            f"Files matching the pattern ``ckpt_dir + ckpt_name`` will be loaded."
+            f"For model, the default pattern is {MODEL_NAME_PATTERN}. For optimizer,"
+            f"the default pattern is {OPTIM_NAME_PATTERN}"),
+    )
+    parser.add_argument(
+        "--ckpt_type",
```
Comment (anw90) [inline torchacc/utils/consolidate_and_reshard_ckpts.py:45]: Could we support consolidating the model and optimizer into one command?

File: torchacc/utils/consolidate_and_reshard_ckpts.py [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+        type=str,
```
Comment (anw90) [inline torchacc/utils/consolidate_and_reshard_ckpts.py:46]: May be we could add a `choices` option for `ckpt_type`.

File: torchacc/utils/consolidate_and_reshard_ckpts.py (hunk 2 above)
Comment (hanwen-sun) [reply torchacc/utils/consolidate_and_reshard_ckpts.py:1]: done. I add the entry_points in setup.py to use consolidate_and_reshape_ckpts after installing torchacc. I‘m not sure if this is what you meant.

File: torchacc/utils/consolidate_and_reshard_ckpts.py (hunk 5 above)
Comment (hanwen-sun) [reply torchacc/utils/consolidate_and_reshard_ckpts.py:45]: Users need to tell us the checkpoint file name they want to load and  consolidate. Like "rank-*of-*-model.pth" or "rank-*of-*-optimizer.pth" .What's more, they need to tell us the dir to save.  We now support users to tell us one name_pattern at once. So they need to consolidate model and optimizer seperately.
We can add args like ``model_ckpt_name``, ``optimizer_ckpt_name``, ``model_save_name``, ``optimizer_save_name`` in the script to support consolidating model and optimizer into one command. 
However, I think may be there are some scenarios that users only want to consolidate model or optimizer?

File: torchacc/utils/consolidate_and_reshard_ckpts.py (hunk 6 above)
Comment (hanwen-sun) [reply torchacc/utils/consolidate_and_reshard_ckpts.py:46]: done

File: torchacc/utils/consolidate_and_reshard_ckpts.py [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+            f"Files will be saved in path: ``save_dir + save_name`.`"
+            f"For consolidated checkpoint, the default path is: ``save_dir + model/optimizer_consolidated.pth``"
+            f"For reshard checkpoints, the default path is: ``save_dir + {DEFULT_MODEL_NAME_PATTERN}/{DEFAULT_OPTIM_NAME_PATTERN}``."
+            f"For reshard checkpoints, please use the same name patthern as {DEFULT_MODEL_NAME_PATTERN} and {DEFAULT_OPTIM_NAME_PATTERN}."
+        ),
+    )
+
+    args = parser.parse_args()
+
+    if args.ckpt_type == "model":
```
Comment (hanwen-sun) [inline torchacc/utils/consolidate_and_reshard_ckpts.py]: Is it suitable to add some judgement and set default value here? Or is it better to remove these code to the consolidate function.

File: torchacc/dist/state_dict_utils.py (hunk 1 above)
Comment (hanwen-sun) [reply torchacc/dist/state_dict_utils.py]: done

File: torchacc/utils/consolidate_and_reshard_ckpts.py (hunk 3 above)
Comment (hanwen-sun) [reply torchacc/utils/consolidate_and_reshard_ckpts.py]: I have renamed them to DEFAULT_MODEL_NAME_PATTERN and DEFAULT_OPTIMIZER_NAME_PATTERN. Users can specify the name patterns using the arguments ckpt_name and save_name. If these two arguments are not specified, the default name patterns will be used. Additionally, these default patterns are used to provide examples of name patterns in the help text for the arguments

File: torchacc/utils/consolidate_and_reshard_ckpts.py (hunk 4 above)
Comment (hanwen-sun) [reply torchacc/utils/consolidate_and_reshard_ckpts.py]: done

File: setup.py [hunk 8]
```diff
@@ -32,6 +32,13 @@ def get_and_set_version():
     packages=['torchacc'] + ['torchacc.' + \
                              pkg for pkg in find_packages('torchacc')],
 
+    # add console_scripts
+    entry_points={
+        'console_scripts': [
+            'consolidate_and_reshape_ckpts = torchacc.utils.consolidate_and_reshard_ckpts:main',
```
Comment (anw90) [inline setup.py]: Rename `consolidate_and_reshape_ckpts` to `consolidate_and_reshape_fsdp_ckpts` (Add fsdp information to the file name).

**Review by anw90 — APPROVED** [review]

**Pull request merged**